References between documents should be displayed as a graph (see #11)

CocoChart opened the pull request:
I, Corentin Chartier, hereby grant to Hyperglosae maintainers the right to publish my contribution under the terms of any licenses the Free Software Foundation classifies as Free Software Licenses.

benel commented:
Thank you very much for your contribution, @CocoChart!

Before going on with the review, I would need you to fix the acceptance test that got broken :

> Running:  bookmark_document.feature                                                       (1 of 7)
> Ajouter un document à sa bibliothèque
>    1) [unamed scenario]
> 0 passing (5s)
>  1 failing
>  1) Ajouter un document à sa bibliothèque
>      [unamed scenario]:
>     CypressError: Timed out retrying after 4050ms: `cy.click()` failed because the page updated while this command was executing. Cypress tried to locate elements based on this query:
> cy.get(**.create-document**)

Source : https://github.com/Hypertopic/HyperGlosae/actions/runs/13525583098/job/37795170633?pr=254#step:10:281

In other words, the test runner cannot locate an element with the `create-document` class.

If you need help, please just ask.

benel commented:
Please note that I updated the documentation on how to run tests on one's desktop: https://github.com/Hypertopic/HyperGlosae/blob/main/CONTRIBUTING.md#running-tests

benel reviewed (changes requested):
- [x] Please note also that there is a borderline case for which it doesn't work: 
<img width="707" alt="Capture d’écran 2025-02-25 à 18 16 23" src="https://github.com/user-attachments/assets/9f8367b4-538f-4c08-b30d-e4047aa4c153" />

When a user has no documents at all, the graph stays on "loading" and the button to create a document from scratch never appears.

Note: Bill is one of the [sample users for tests](https://github.com/Hypertopic/HyperGlosae/tree/main/samples/_users).

CocoChart commented:
> Before going on with the review, I would need you to fix the acceptance test that got broken :
> 
> > Running:  bookmark_document.feature                                                       (1 of 7)

The test fails on GitHub but passes when run with Cypress on my desktop.

![image](https://github.com/user-attachments/assets/7b6424d8-2416-4e0e-9b00-d56825a3457d)

benel commented:
@CocoChart 

> The test fails on GitHub but passes when run with Cypress on my desktop.

OK... It means probably that loading the graph takes too much time on CI.

- [ ] Please try to display the `.create-document` button independently from the graph loading (from a functional side, there is no reason to differ displaying this button).

CocoChart commented:
The create document button is now displayed independently of the graph loading. The tests all work locally, waiting to see if they work on GitHub

benel commented:
You got again the integration problem I resolved yesterday (waiting endlessly for "work").
This is because you modified your commit and not the one I updated.

So that you are more independent please: 

1. first create a backup branch on your commit (in case you would do something else that what you should),
2. go to the main branch and pull the last version,
3. check your branch (`requirement-11`),
4. rebase it on main,
5. fix any conflicts and go on with rebasing,
6. test that everything is OK (install and start),
7. force push to your remote branch.

CocoChart commented:
Sorry about that. I corrected the problem by rebasing my branch on `main` and updated my code again.

benel commented:
> All checks have passed.

Cool! 

I will now go on with the review process.

benel commented on frontend/src/routes/Bookshelf.jsx:
```diff
@@ -2,22 +2,23 @@ import '../styles/Bookshelf.css';
+import GraphContainer from '../components/GraphContainer';
 
 function Bookshelf({backend, user}) {
+  console.log('Bookshelf');
```
Console logs are often necessary at development time but they should be cleaned away at integration time so that future developers won't get spammed.

benel commented on frontend/src/routes/Bookshelf.jsx:
```diff
@@ -2,22 +2,23 @@ import '../styles/Bookshelf.css';
   const [lastUpdate, setLastUpdate] = useState();
 
   useEffect(() => {
+    console.log('Fetching documents...');
```
Idem.

benel commented on frontend/src/routes/Bookshelf.jsx:
```diff
@@ -2,22 +2,23 @@ import '../styles/Bookshelf.css';
-        {...{setLastUpdate, backend}}
-      />
-    </Container>
+    <>
```
An empty elements is necessary in JSX only when the HTML fragment contains more than one root element.

benel commented on frontend/src/components/GraphContainer.jsx:
```diff
@@ -0,0 +1,27 @@
+  const displayedDocs = docs.flatMap(d => d[0]);
+
+  return (
+    <>
```
An empty elements is necessary in JSX only when the HTML fragment contains more than one root element.

benel reviewed (changes requested):
Thank you very much for this contribution @CocoChart!

You will find below some required improvements.

Moreover, please consider that "[comments in code are a bad thing](https://www.quora.com/Are-comments-in-code-a-good-or-bad-thing)".

- [ ] For your code, the only part that may be documented inline is perhaps line 133-134 because `M` and `A` are not self explanatory. But instead of a long comment you may refer to the Web page where this syntax is defined.

Thank you in advance. I'm eager to push your contribution to the production server.

benel commented:
@CocoChart 
One last thing, please change the current commit message from
`REQUIREMENT : References between documents should be displayed as a graph (closes #11)` to 
`REQUIREMENT: References between documents should be displayed as a graph (closes #11).` .

✍🏼

CocoChart commented:
The code has now been improved as requested. For lines 133-134, I added a link to the MDN documentation about SVG paths. I also renamed some variables for better clarity.

benel commented:
Thank you very much @CocoChart! We are really near a new release.
The last problem to solve is this one: https://github.com/Hypertopic/HyperGlosae/pull/254#pullrequestreview-2641896319

This is a real problem since it means that on the production server "Loading..." would be displayed forever (until someone logs in).

benel commented:
@CocoChart 
Please note that the way to refresh user documents has changed.

```
backend.refreshDocuments(
  x => setUserDocuments(f)
);
```
is replaced by
```
backend.getAllDocuments(user)
  .then(x =>
    setUserDocuments(f)
  );
```

It means that you have to pass `user` as a parameter from `App.jsx` to the component that refreshes documents.

If everything goes right, your graph should reload when a user logs out.

benel commented:
Thank you @CocoChart!
I will look at this ASAP.

benel commented on frontend/package.json:
```diff
@@ -5,6 +5,9 @@
   "type": "module",
   "dependencies": {
     "bootstrap": "^5.2.2",
+    "buffer": "^6.0.3",
```
`buffer` package is no more needed. It has been removed in the main branch.

benel commented on frontend/src/App.jsx:
```diff
@@ -35,7 +35,7 @@ function App() {
         <TypesContext.Provider value={types}>
           <Routes>
-            <Route path="/" element={<Bookshelf {...{backend, user}} />} />
+            <Route path="/" element={<Bookshelf createOn={[]} {...{backend, user}} />} />
```
`createOn` is just an empty list. And `Bookshelf` is used nowhere else. So there is no reason to pass it as a parameter.

benel commented on frontend/src/components/Graph.jsx:
```diff
@@ -0,0 +1,158 @@
+    const typeColorScale = scaleOrdinal(types, schemeCategory10);
+    const nodes = docs.map(d => {
+      return { id: d[0], title: d[1] };
+    });
```
Don't use return when you have a single line function:

```
d => ({ id: d[0], title: d[1] })
```

The enclosing parenthesis is here so that `{}` are interpreted as enclosing an object and not a code block.

benel replied:
Or you can even replace it with the more elegant:

```
[id, title] => ({id, title})
```

benel commented on frontend/src/routes/Bookshelf.jsx:
```diff
@@ -1,23 +1,34 @@
 
-function Bookshelf({backend, user}) {
-  const [documents, setDocuments] = useState([]);
+function Bookshelf({ createOn, backend, user }) {
```
`createOn` is always an empty list. So there is no reason to pass it as a parameter.

benel commented on frontend/src/routes/Bookshelf.jsx:
```diff
@@ -1,23 +1,34 @@
-      />
+      <Row>
+        <Col className="col-10">
+          <Graph rawDocs={docs} displayedDocs={displayedDocs} />
```
For more coherence you may write:

    {..{displayedDocs}}

instead of:

    displayedDocs={displayedDocs}

It is equivalent however.

benel commented on frontend/src/routes/Bookshelf.jsx:
```diff
@@ -1,23 +1,34 @@
+        </Col>
+        {createOn &&
+      <Col>
+        <FutureDocument relatedTo={createOn} {...{setLastUpdate, backend, user}} />
```
`createOn` is always an empty list. So there is no reason have it as a parameter.

You could write:

    <FutureDocument relatedTo={[]} {...{setLastUpdate, backend, user}} />

benel commented on frontend/src/routes/Bookshelf.jsx:
```diff
@@ -1,23 +1,34 @@
+        {createOn &&
+      <Col>
+        <FutureDocument relatedTo={createOn} {...{setLastUpdate, backend, user}} />
+      </Col>
```
Be careful, lines 27-29 should be indented by 4 more spaces.

benel reviewed (changes requested):
Thank you very much. A few last things...

benel reviewed (approved):
Yahoo!!! Well done!
